src/pool/auction.ts

```diff
@@ -0,0 +1,254 @@
+    }
+    return new ScaledAuction(
+      this.user,
+      this.type,
+      scaledAuction.bid,
+      scaledAuction.lot,
+      this.auctionData.block,
+      scaleBlock,
+      filled
+    );
+  }
+}
+
+export class ScaledAuction {
+  type: AuctionType;
+  user: string;
+  bid: Map<string, i128>;
+  lot: Map<string, i128>;
+  startBlock: number;
+  scaleBlock: number;
+  filled: boolean;
+  // the hash of the transaction that filled the auction
+  hash: string | undefined;
+  // the timestamp of the auction creation
+  timestamp: string | undefined;
+
+  constructor(
+    user: string,
+    type: AuctionType,
+    bid: Map<string, i128>,
+    lot: Map<string, i128>,
+    startBlock: number,
+    scaleBlock: number,
+    filled: boolean,
+    hash?: string,
+    timestamp?: string
+  ) {
+    this.user = user;
+    this.type = type;
+    this.bid = bid;
+    this.lot = lot;
+    this.startBlock = startBlock;
+    this.scaleBlock = scaleBlock;
+    this.filled = filled;
+    this.hash = hash;
+    this.timestamp = timestamp;
+  }
+
+  /**
+   * Creates an array of scaled auctions from events representing the state of the auctions at the current ledger
+   * @param events - The events to create the scaled auctions from
+   * @param backstopId - The address of the backstop
+   * @param currLedger - The current ledger
+   * @returns An array of scaled auctions
+   */
+  static fromEvents(events: PoolEvent[], backstopId: string, currLedger: number): ScaledAuction[] {
+    const auctions: Auction[] = [];
+    const scaledAuctions: ScaledAuction[] = [];
+    for (const event of events ?? []) {
+      switch (event.eventType) {
+        case PoolEventType.NewLiquidationAuction:
+          auctions.unshift(
+            new Auction(
+              event.user,
+              AuctionType.Liquidation,
+              event.auctionData,
+              event.ledgerClosedAt
+            )
+          );
+          break;
+        case PoolEventType.NewAuction:
+          auctions.unshift(
+            new Auction(backstopId, event.auctionType, event.auctionData, event.ledgerClosedAt)
+          );
+          break;
+        case PoolEventType.FillAuction: {
+          const index = auctions.findIndex(
+            (auction) => auction.user === event.user && auction.type === event.auctionType
+          );
+          if (index !== -1) {
+            const auction = auctions[index];
+            let scaledAuction: ScaledAuction;
+            if (event.fillAmount < BigInt(100)) {
+              scaledAuction = auction.scale(event.ledger, Number(event.fillAmount), true);
+              // Scale the auction to start block + 200 to get full auction
+              const remainingAuction = auction.scale(
+                auction.auctionData.block + 200,
+                100 - Number(event.fillAmount),
+                false
+              );
+              auction.auctionData.lot = remainingAuction.lot;
+              auction.auctionData.bid = remainingAuction.bid;
```

Comment:
on a partial auction fill, we should store a new auction marked as `filled` containing the scaled auction, and update the existing auction to the remaining auction, where the remaining auction is the delta from the base auction to the scaled auction.